src/gadgets/k_mix.rs

```diff
@@ -1,61 +1,213 @@
 #![allow(non_snake_case)]
 
 use super::mix;
-use bulletproofs::r1cs::ConstraintSystem;
-use error::SpacesuitError;
+use bulletproofs::r1cs::{ConstraintSystem, R1CSError};
+use curve25519_dalek::scalar::Scalar;
 use std::iter::once;
-use value::AllocatedValue;
+use subtle::{ConditionallySelectable, ConstantTimeEq};
+use value::{AllocatedValue, Value};
 
 /// Enforces that the outputs are either a merge of the inputs: `D = A + B && C = 0`,
 /// or the outputs are equal to the inputs `C = A && D = B`. See spec for more details.
 /// Works for `k` inputs and `k` outputs.
 pub fn fill_cs<CS: ConstraintSystem>(
     cs: &mut CS,
     inputs: Vec<AllocatedValue>,
-    intermediates: Vec<AllocatedValue>,
-    outputs: Vec<AllocatedValue>,
-) -> Result<(), SpacesuitError> {
+) -> Result<(Vec<AllocatedValue>, Vec<AllocatedValue>), R1CSError> {
     // If there is only one input and output, just constrain the input
     // and output to be equal to each other.
-    if inputs.len() == 1 && outputs.len() == 1 {
+    if inputs.len() == 1 {
         let i = inputs[0];
-        let o = outputs[0];
+        let o = i.reallocate(cs)?;
         cs.constrain(i.q - o.q);
         cs.constrain(i.a - o.a);
         cs.constrain(i.t - o.t);
-        return Ok(());
+        return Ok((vec![i], vec![o]));
     }
 
-    if inputs.len() != outputs.len() || intermediates.len() != (inputs.len() - 2) {
-        return Err(SpacesuitError::InvalidR1CSConstruction);
-    }
+    let (mix_in, mix_mid, mix_out) = make_intermediate_values(&inputs, cs)?;
+    call_mix_gadget(cs, &mix_in, &mix_mid, &mix_out)?;
+    Ok((mix_in, mix_out))
+}
 
-    let first_input = inputs[0].clone();
-    let last_output = outputs[outputs.len() - 1].clone();
+fn call_mix_gadget<CS: ConstraintSystem>(
+    cs: &mut CS,
+    mix_in: &Vec<AllocatedValue>,
+    mix_mid: &Vec<AllocatedValue>,
+    mix_out: &Vec<AllocatedValue>,
+) -> Result<(), R1CSError> {
+    let first_in = mix_in[0].clone();
+    let last_out = mix_out[mix_out.len() - 1].clone();
 
     // For each 2-mix, constrain A, B, C, D:
     for (((A, B), C), D) in
-        // A = (first_input||intermediates)[i]
-        once(first_input).chain(intermediates.clone().into_iter())
-        // B = inputs[i+1]
-        .zip(inputs.into_iter().skip(1))
-        // C = outputs[i]
-        .zip(outputs.into_iter())
-        // D = (intermediates||last_output)[i]
-        .zip(intermediates.into_iter().chain(once(last_output)))
+        // A = (first_in||mix_mid)[i]
+        once(&first_in).chain(mix_mid.iter())
+        // B = mix_in[i+1]
+        .zip(mix_in.iter().skip(1))
+        // C = mix_out[i]
+        .zip(mix_out.iter())
```

Comment:
We have K input/output wires with K-2 middle wires.

The loop is over K-1 blocks of 2x2 mix wires, but `.zip(mix_out.iter())` takes K output wires and silently truncates it to K-1 because all other iterators are of K-1 length. This is a documented behavior of `zip`, but I don't quite like it and it means, the correctness of this line alone is now dependent on the correctness of the other lines. We should probably do `.zip(mix_out.iter().take(mix_out.len()-1))` to make this explicit.

Also would help to document the loop: what lengths are of all the arrays and what is the length of the loop itself.

Reply:
(would also help to do `let k = mix_out.len()` and use `k` for clarity)